lib/workers/repository/update/pr/code-owners.ts

```diff
@@ -21,37 +32,129 @@ export async function codeOwnersForPr(pr: Pr): Promise<string[]> {
 
     logger.debug(`Found CODEOWNERS file: ${codeOwnersFile}`);
 
+    // Get list of modified files in PR
     const prFiles = await getBranchFiles(pr.sourceBranch);
-    const rules = codeOwnersFile
+
+    if (!prFiles) {
+      logger.debug('PR includes no files');
+      return [];
+    }
+
+    // Convert CODEOWNERS file into list of matching rules
+    const fileOwnerRules = codeOwnersFile
       .split(newlineRegex)
+      // Remove empty and commented lines
       .map((line) => line.trim())
       .filter((line) => line && !line.startsWith('#'))
+
+      // Extract pattern & usernames
       .map((line) => {
         const [pattern, ...usernames] = line.split(regEx(/\s+/));
         return {
           usernames,
-          match: (path: string) => {
-            const matcher = ignore().add(pattern);
-            return matcher.ignores(path);
-          },
+          pattern,
+          match: (path: string) => ignore().add(pattern).ignores(path),
         };
-      })
-      .reverse();
+      });
+
     logger.debug(
-      { prFiles, rules },
+      { prFiles, fileOwnerRules },
       'PR files and rules to match for CODEOWNERS'
     );
-    const matchingRule = rules.find((rule) => prFiles?.every(rule.match));
-    if (!matchingRule) {
-      logger.debug('No matching CODEOWNERS rule found');
-      return [];
-    }
+
+    // Apply rules & get list of owners for each prFile
+    const emptyRules = fileOwnerRules.filter(
+      (rule) => rule.usernames.length === 0
+    );
+    const fileOwners =
+      // Map through all prFiles and match said file with all the rules
+      prFiles
+        .map<{
+          file: string;
+          usernames: UserScore[];
+        }>((file) => {
+          const usersWithScore = fileOwnerRules
+            .map((rule) =>
+              rule.match(file)
+                ? { score: rule.pattern.length, usernames: rule.usernames }
+                : null
+            )
+            .reduce<UserScore[]>((acc, match) => {
+              if (!match) {
+                return acc;
+              }
+              for (const user of match.usernames) {
+                if (!acc.find((a) => a.username === user)) {
+                  acc.push({ score: match.score, username: user });
+                }
+              }
+
+              return acc;
+            }, []);
+
+          return { file, usernames: usersWithScore };
+        })
+
+        // Match file again but this time only with emptyRules, this makes sure that files that have been set as having no owner remain as such
```

Comment:
```suggestion
        // Match file again but this time only with emptyRules, to ensure that files which have no owner set remain owner-less
```

I think this is easier to understand, but check if it's still technically correct. 😉